### PR TITLE
[k2 integration] New PR copied from #2133

### PR DESCRIPTION
**Pull request by ZhaoZeyu1995**
## What does this PR do?

Hello,

This is a copied PR from [the old one](https://github.com/speechbrain/speechbrain/pull/2133). Please refer to the old PR for more details.

### CODE REVIEW

**Comment by Adel-Moumen** [comment]
Thanks guys for the PR. 

So this is my bucket to do list: 
1. train every models and release them on our dropbox 
2. finalise DDP testing
3. finale review + fix some little things 

I'll get back to you when I did everything :)

File: recipes/LibriSpeech/ASR/CTC/README.md [hunk 1]
```diff
@@ -57,6 +78,8 @@ Note: by default, `topk` is set to 20 as it gives a good trade-off between WER a
...
 | 08-12-23 | train_hf_whisper.yaml (small) | CTCBeamSearch + 4-gram  + test batch size = 1 | 960h | 4.37 | 3.16 | 11.76 | 3.43 | Not Avail. | [Link](https://www.dropbox.com/sh/zmtp13huxn02fot/AADyKL5q0MwRhEG1-WbSXDWda?dl=0) |  1xRTX3090 24GB | 2xTesla V100 32GB |
 | 08-12-23 | train_hf_whisper.yaml (small) | CTCPrefixBeamSearch + 4-gram  + test batch size = 1 | 960h | 4.44 | 3.30 | 11.89 | 3.47 | Not Avail. | [Link](https://www.dropbox.com/sh/zmtp13huxn02fot/AADyKL5q0MwRhEG1-WbSXDWda?dl=0) |  1xRTX3090 24GB | 2xTesla V100 32GB |
+| 10-05-23 | train_hf_wav2vec_k2.yaml | k2CTC + HL graph + 1best decoding + test batch size = 1 | 960h | 1.84 | Not Avail. | 3.81 | Not Avail. | Not Avail. | Not Avail. |  1xRTX2080Ti 12GB | 1xRTX2080Ti 12GB |
+| 10-05-23 | train_hf_wav2vec_k2.yaml | k2CTC + HLG graph + 1best decoding + test batch size = 1 | 960h | 1.72 | Not Avail. | 3.46 | Not Avail. | Not Avail. | Not Avail. |  1xRTX2080Ti 12GB | 1xRTX2080Ti 12GB |
```
Comment (Adel-Moumen) [inline recipes/LibriSpeech/ASR/CTC/README.md:82]: I will update the path to the dropbox shortly I just need to rerun training exp

File: speechbrain/k2_integration/lexicon.py [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+                raise RuntimeError(
+                    f"{self.lang_dir}/L_disambig.pt does not exist. Please make sure "
+                    f"you have successfully created L_disambig.pt in {self.lang_dir}"
+                )
+        return self._L_disambig
+
+    def remove_G_rescoring_disambig_symbols(self, G: k2.Fsa):
+        """
+        Remove the disambiguation symbols of a G graph
+        """
```
Comment (Adel-Moumen) [inline speechbrain/k2_integration/lexicon.py:155]: missing docstring arg for G

File: speechbrain/k2_integration/losses.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    is_training=True,
+):
+    """
+    CTC loss implemented with k2. Make sure that k2 has been installed properly.
+    Note that the blank index must be 0 in this implementation.
+
+    Arguments
+    ---------
+    predictions : torch.Tensor
+        Predicted tensor, of shape [batch, time, chars].
```
Comment (Adel-Moumen) [inline speechbrain/k2_integration/losses.py:32]: I suppose you meant log_probs ?

File: speechbrain/k2_integration/prepare_lang.py [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+    """
+    Get tokens from a lexicon.
+
+    Arguments
+    ---------
+    lexicon: Lexicon
+        It is the return value of :func:`read_lexicon`.
+    sil_token: str
+        The optional silence token between words. It should not appear in the lexicon,
+        otherwise it will cause an error.
```
Comment (Adel-Moumen) [inline speechbrain/k2_integration/prepare_lang.py:60]: missing docstring arg for ` manually_add_sil_to_tokens=False`

File: speechbrain/k2_integration/utils.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+def lattice_paths_to_text(best_paths: k2.Fsa, word_table) -> List[str]:
+    """
+    Convert the best path to a list of strings.
+
+    Arguments
+    ---------
+    best_paths: k2.Fsa
+        It is the path in the lattice with the highest score for a
+        given utterance.
```
Comment (Adel-Moumen) [inline speechbrain/k2_integration/utils.py:77]: missing word_table from docstring

File: speechbrain/k2_integration/utils.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    """
+    Prepare a LM with the purpose of using it for LM rescoring.
+    For instance, in the librispeech recipe this is a 4-gram LM (while a
+    3gram LM is used for HLG construction).
+
+    Arguments
+    ---------
+    path: str
+        The path to an FST LM (ending with .fst.txt) or a k2-converted
+        LM (in pytorch .pt format).
```
Comment (Adel-Moumen) [inline speechbrain/k2_integration/utils.py:145]: G should be there and you should remove path as it is not part of the arguments of this function

File: speechbrain/k2_integration/graph_compiler.py [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+        Compile the decoding graph by composing H with LG.
+        This is for decoding with small language model.
+
+        Arguments
+        ---------
+        cache_dir: str
+            The path to store the composition in a .pt format.
+        cache: bool
+            Whether or not to load the composition from the .pt format (in the
+            cache_dir dir).
```
[the remark this answers is not in the document]
Comment (ZhaoZeyu1995) [reply speechbrain/k2_integration/graph_compiler.py]: Solved.

File: speechbrain/k2_integration/lattice_decoder.py [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+        Set it to zero if there is no constraint.
+    max_active_states:
+        Maximum number of FSA states that are allowed to be active on any given
+        frame for any given intersection/composition task. This is advisory,
+        in that it will try not to exceed that but may not always succeed.
+        You can use a very large number if no constraint is needed.
+    ac_scale:
+        acoustic scale applied to `log_probs_nnet_output`
+    subsampling_factor:
+        The subsampling factor of the model.
```
[the remark this answers is not in the document]
Comment (ZhaoZeyu1995) [reply speechbrain/k2_integration/lattice_decoder.py]: Solved.

File: speechbrain/k2_integration/lexicon.py [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+        The directory to store the lexicon.txt
+    vocab_files: List[str]
+        A list of extra vocab files. For example, for librispeech this could be the
+        librispeech-vocab.txt file.
+    extra_csv_files: List[str]
+        A list of csv file paths
+    add_word_boundary: bool
+        whether to add word boundary symbols <eow> at the end of each line to the
+        lexicon for every word.
+
```
[the remark this answers is not in the document]
Comment (ZhaoZeyu1995) [reply speechbrain/k2_integration/lexicon.py]: Solved.

File: speechbrain/k2_integration/lexicon.py (hunk 2 above)
Comment (ZhaoZeyu1995) [reply speechbrain/k2_integration/lexicon.py]: Solved.

File: speechbrain/k2_integration/utils.py (hunk 5 above)
Comment (ZhaoZeyu1995) [reply speechbrain/k2_integration/utils.py:77]: Solved.

File: speechbrain/k2_integration/utils.py (hunk 6 above)
Comment (ZhaoZeyu1995) [reply speechbrain/k2_integration/utils.py]: Solved.

**Comment by ZhaoZeyu1995** [comment]
Hello @Adel-Moumen,

I just added the docstrings you required. Please feel free to let me know if there are any other issues. 

Thank you very much!

**Comment by Adel-Moumen** [comment]
> Hello @Adel-Moumen,
> 
> I just added the docstrings you required. Please feel free to let me know if there are any other issues.
> 
> Thank you very much!

Thanks! 

I'll ping @mravanelli and @TParcollet so that they can have a look at the PR. 

On my side I was able to reproduce the numbers, and we have now a clear separation between training a wav2vec2 and ngram creation. 

I am finalising the experiments and uploading them on dropbox. I tried with DDP/single gpu and everything works as expected.

**Comment by Adel-Moumen** [comment]
K2 recipe test passed. 
```
(1/1) Running test for LibriSpeech_row_44...
        ... 94.23s
TEST PASSED
```

**Comment by Adel-Moumen** [comment]
Test passed as well for 1best + HLG
```
(1/1) Running test for LibriSpeech_row_44...
        ... 1192.45s
TEST PASSED
```

I'm wondering though if we should have one yaml for each different models e.g. 1best + HLG/HL, whole-lattice-decoding + HLG/HL, etc.

**Comment by Adel-Moumen** [comment]
I added the final results. Now we should be ready to go. Ping @TParcollet and @mravanelli for a final check.

**Comment by ZhaoZeyu1995** [comment]
> Test passed as well for 1best + HLG
> 
> ```
> (1/1) Running test for LibriSpeech_row_44...
>         ... 1192.45s
> TEST PASSED
> ```
> 
> I'm wondering though if we should have one yaml for each different models e.g. 1best + HLG/HL, whole-lattice-decoding + HLG/HL, etc.

I would say we may just have one yaml file (1best+HLG), and possibly mention in the README or a tutorial that there are several different combinations of decoding settings.

**Review by mravanelli — APPROVED** [review]

**Comment by mravanelli** [comment]
I did my last review and I ran some tests. Everything seems to work fine. Let me thank all the contributors to this PR. This is an amazing work and an important milestone for SpeechBrain!

**Comment by Adel-Moumen** [comment]
Thanks @ZhaoZeyu1995 @pchampio @geoph9 for doing such an amazing work. Congrats for your first contribution (and big). Thanks also @csukuangfj for helping a bit! (btw we are looking for the torch 2.2 k2 wheels :eyes:)

**Comment by csukuangfj** [comment]
> Thanks @ZhaoZeyu1995 @pchampio @geoph9 for doing such an amazing work. Congrats for your first contribution (and a big one!). Thanks also @csukuangfj for helping a bit! (btw we are looking for the torch 2.2 k2 wheels :eyes:)

please see our latest doc. We provide both CPU and CUDA pre-built k2 wheels for torch 2.2.0

**Pull request merged**